src/txn/txn.c

```diff
@@ -210,9 +210,23 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
 		 *    an ID -- the ID will not be used because the thread will
 		 *    keep spinning until it gets a valid one.
 		 */
-		if (s != txn_state &&
+retry:		if (s != txn_state &&
 		    (id = s->id) != WT_TXN_NONE &&
 		    WT_TXNID_LE(prev_oldest_id, id)) {
+			/*
+			 * If the transaction is still allocating its ID,
+			 * then we spin here until it gets its valid ID.
+			 * There might be a chance that this ID can become
+			 * higher if the transaction is quickly done its job.
+			 * In this case, we ignore this transaction since
+			 * it would not be visible anyway in current snapshot.
+			 */
+			if (WT_TXNID_LT(current_id, id))
+				continue;
+			if (s->is_allocating != false) {
+				WT_PAUSE();
+				goto retry;
+			}
```

Comment:
This code can make an incorrect state if the execution order is
1. Thread A (txn allocating its txn id) set its `is_allocating` value as true and set its id with the `current` value. (up to __wt_txn_id_alloc:992)
2. Thread B (txn creating its snapshot) check if the Thread A's txn id is smaller than its own id. (up to __wt_txn_get_snapshot:225)
3. Thread A set its real txn id and turn its `is_allocating` flag as false. (up to __wt_txn_id_alloc:995)
4. Thread B sees the `is_allocating` flag as false (line 226) and then store the id as Thread A's txn id, which is different from the real B's ID.

I think the root cause of this problem is that the `is_allocating` flag and the id value are decoupled into different variables so any thread can see the middle state of it. IMO, it can be simpler to just steal 1 bit of the id variable and use it for the is_allocating flag so we can easily & atomically read the transaction's state. One thing I'm worried about is that this bit stealing can reduce the flexibility of the code so I think it's better to get confirm from another expert.

Reply:
To the best of my knowledge, I resolved this bug by checking `id == s->id` again after seeing `is_allocating` as false. If once fetched `id` is not the same with current `s->id`, then this `id` might not be a valid one and by iterating loop it will catch the valid one (or ignore if this transaction already completed its job or become higher). I'm not sure where to put the correct read barrier in this changed code yet and it requires discussion.

Reply:
I agree that a read barrier is required in between checking `is_allocating` and re-checking `id`.  Avoiding this complexity is definitely a point in favour of stealing a bit from the transaction ID.

Reply:
@michaelcahill, I also added another read barrier at line 225 to prevent reordering between reading `id` (at line 218) and reading `is_allocating` (at line 226) because this order should not be changed too.
@ZombieKim and I discussed and we're not sure if this kind of reordering could occur. Because we think `id` and `is_allocating` have no dependency logically so this might happen, though this two variables would be in the same cache line. Do you think this is necessary?

Reply:
I wouldn't make assumptions based on cache lines: WT runs on multiple CPU architectures.

At this point, assuming there is no performance regression on any of our workloads, I am happy with the clarity of having two read barriers.  I suspect we can rewrite this code to read `is_allocating` first, so it has a single read barrier, but I don't think that should hold up getting this change merged.

Reply:
> I suspect we can rewrite this code to read `is_allocating` first, so it has a single read barrier

I think this way can make another problem. If `is_allocating` value has read before setting to `true` and then `id` value has read as a temporal id value which could be different from real txn id value, snapshot can miss the real txn id.